### PR TITLE
Replace pointer casts in std.math with union repainting

### PR DESCRIPTION
**Pull request by ivan-timokhin**
From the comments received on #3022, I was left with an impression that unions are generally preferred to pointer casts, and should be used instead wherever possible. So I went ahead and replaced most of the pointer casts with unions.

This is mostly straightforward one-to-one replacement, with notable exception of `isIdentical`, because I couldn't resist the temptation.

I didn't touch `NaN` and `getNaNPayload`, because
1. I couldn't figure out what real formats are actually supported by them.
2. I've no idea how to translate these weird unaligned reads/stores [[1]](https://github.com/D-Programming-Language/phobos/blob/master/std/math.d#L5257-L5264)[[2]](https://github.com/D-Programming-Language/phobos/blob/master/std/math.d#L5309-L5316) into unions in a sane way.

Destroy!

### CODE REVIEW

**Comment by ibuclaw** [comment]
Let's start with the bike shedding first, as that should be the easy bit.... :-)
- ``Repainter``: I'd be leaning on something like ``FloatLayout`` to describe this, but maybe it is better suited for...
-  ``Blob``: Because when I think of blob, object storage systems come to mind.

> I didn't touch `NaN` and `getNaNPayload`, because
> 1. I couldn't figure out what real formats are actually supported by them.

64bit, 128bit, and x87 reals by the looks of it.  (Considering that DMD code would be using the catch-all else block in there).

There should be an explicit ``static if`` condition and ``else static assert``.

> 2. I've no idea how to translate these weird unaligned reads/stores [[1]](https://github.com/D-Programming-Language/phobos/blob/master/std/math.d#L5257-L5264)[[2]](https://github.com/D-Programming-Language/phobos/blob/master/std/math.d#L5309-L5316) into unions in a sane way.

You'll just be dereferencing the ``.bytes`` array I guess, but the cast from ``ubyte*`` to ``ulong*`` is interesting.

Maybe from ``*cast(ulong*)(6+cast(ubyte*)(&x))`` to ``*cast(ulong*)(&x.bytes[6])``.  Not that I approve of such magic.  Which leads back to an earlier point of leveraging some ``FloatLayout`` structure into the union.

I'm thinking along the lines of:
```
static if (T.mant_dig == 24)
{
  version(BigEndian)
    alias Layout = Tuple!("negative", ubyte, "exponent", ulong, "qnan", ubyte, "mantissa", ubyte[22]);
  else
    alias Layout = Tuple!("mantissa", ubyte[22], "qnan", ubyte, "exponent", ulong, "negative", ubyte);
}
```
Something that you have already sort of done with ``significand`` using x87 reals.

**Comment by ivan-timokhin** [comment]
On Sat, Jul 18, 2015 at 05:07:57AM -0700, Iain Buclaw wrote:
> Let's start with the bike shedding first, as that should be the easy bit.... :-)
> - ``Repainter``: I'd be leaning on something like ``FloatLayout`` to describe this, but maybe it is better suited for...
> -  ``Blob``: Because when I think of blob, object storage systems come to mind.
> 

Well, I don't like the names too much myself :). ``FloatLayout`` is probably
even better suited for what you describe below, so...

> 
> I'm thinking along the lines of:
> ```
> static if (T.mant_dig == 24)
> {
>   version(BigEndian)
>     alias Layout = Tuple!("negative", ubyte, "exponent", ulong, "qnan", ubyte, "mantissa", ubyte[22]);
>   else
>     alias Layout = Tuple!("mantissa", ubyte[22], "qnan", ubyte, "exponent", ulong, "negative", ubyte);
> }
> ```
> Something that you have already sort of done with ``significand`` using x87 reals.
> 

I can probably try and come up with something using bitfields, but I'd like to
limit the scope of this pull request to simple translation, and add further
improvements later.

**Comment by 9il** [comment]
Great work! Thanks!

> I can probably try and come up with something using bitfields, but I'd like to
> limit the scope of this pull request to simple translation, and add further
> improvements later.

+1

**Comment by ivan-timokhin** [comment]
Actually, it looks like we have some interesting stuff in Phobos: [FloatRep](http://dlang.org/phobos/std_bitmanip.html#.FloatRep), [DoubleRep](http://dlang.org/phobos/std_bitmanip.html#.DoubleRep). And they've been there for a *long* time.

**Comment by 9il** [comment]
FloatRep, DoubleRep is not fast enough.

**Comment by ivan-timokhin** [comment]
But they *do* look similar to what Iain suggested, so maybe a better course of action would be to improve them?

**Comment by 9il** [comment]
Any improvement will break code. The *bitfields* concepts is slow for math.

**Comment by ibuclaw** [comment]
Yes, ``std.bitmanip`` generates some pretty funky accessors, and the existing representations don't consider endian-ness.

**Comment by ivan-timokhin** [comment]
Hm. Then I don't quite understand what the proposed `FloatLayout` is supposed to be.

**Comment by ivan-timokhin** [comment]
Anyhow, what I really wanted to know was whether I should try and unify the code with `(Float|Double)Rep`. The answer's negative, so I suggest we return to the discussion of the PR at hand.

**Comment by ZombineDev** [comment]
@ivan-timokhin perhaps after this PR is done we can make `(Float|Double)Rep` forward to it?

**Comment by 9il** [comment]
> @ivan-timokhin perhaps after this PR is done we can make (Float|Double)Rep forward to it?

FloatTraits is package stuff

**Comment by ivan-timokhin** [comment]
Uhh, no. Any bitfieldish stuff like the one done in those two is most definitely out of scope of this PR. I brought them up because if we *do* want something similar in `floatTraits` in the long-term, it might be better to build on existing solution instead of rolling out a completely new one. Turns out that's a bad idea.

**Comment by quickfur** [comment]
Will the result be CTFE-able?

**Comment by 9il** [comment]
@quickfur Currently unions are not CTFE-able. But they can be in the future.

**Comment by ibuclaw** [comment]
> @quickfur Currently unions are not CTFE-able. But they can be in the future.

This is on my long todo list.

**Comment by ivan-timokhin** [comment]
@ibuclaw I've tried to come up with better names.

**Comment by ibuclaw** [comment]
@ivan-timokhin - OK.  I've been out of action this week, I'll have a review when I can.

In any case, this is for post-2.068 release.

**Comment by DmitryOlshansky** [comment]
@ibuclaw any chance to review/destroy this ? ;)

**Comment by ivan-timokhin** [comment]
I've rewritten functions already using unions for uniformity, and the rewrite is usually very straightforward (though I've taken some liberties in `cmp`), but I can drop these changes to keep this PR focused on getting rid of pointers. Should I?

File: std/math.d [hunk 1]
```diff
@@ -307,6 +311,9 @@ template floatTraits(T)
                 enum EXPPOS_SHORT = 0;
                 enum SIGNPOS_BYTE = 0;
             }
+
+            import std.typecons : Tuple;
+            alias Integral = Tuple!(ulong, ushort);
```
Comment (ibuclaw) [inline std/math.d]: This should be the same as ``mant_dig == 64``.  All bit manipulation is identical to ``ieeeExtended`` for this type.

File: std/math.d [hunk 2]
```diff
@@ -328,6 +335,9 @@ template floatTraits(T)
             enum EXPPOS_SHORT = 0;
             enum SIGNPOS_BYTE = 0;
         }
+
+        import std.typecons : Tuple;
+        alias Integral = Tuple!(ulong, "significand", ushort);
```
Comment (ibuclaw) [inline std/math.d]: What about ``LittleEndian``?  (Where's an ``m68k`` or ``i64`` emulator when you need one?)

File: std/math.d [hunk 3]
```diff
@@ -4092,37 +4105,40 @@ long lrint(real x) @trusted pure nothrow @nogc
...
                     if (exp < 0)
                         result = 0;
                     else if (exp <= 31)
-                        result = vi[1] >> (31 - exp);
+                        result = rep.ints[1] >> (31 - exp);
                     else
-                        result = (cast(long) vi[1] << (exp - 31)) | (vi[0] >> (63 - exp));
+                        result =
+                            (cast(long) rep.ints[1] << (exp - 31)) |
+                            (rep.ints[0] >> (63 - exp));
```
Comment (ibuclaw) [inline std/math.d]: I don't think that splitting it across three lines makes it any more readable.

File: std/math.d [hunk 4]
```diff
@@ -4092,37 +4105,40 @@ long lrint(real x) @trusted pure nothrow @nogc
...
                     if (exp < 0)
                         result = 0;
                     else if (exp <= 31)
-                        result = vi[1] >> (31 - exp);
+                        result = rep.ints[1] >> (31 - exp);
                     else
-                        result = (cast(long) vi[1] << (exp - 31)) | (vi[2] >> (63 - exp));
+                        result =
+                            (cast(long) rep.ints[1] << (exp - 31)) |
+                            (rep.ints[2] >> (63 - exp));
```
Comment (ibuclaw) [inline std/math.d]: Same here, I wouldn't worry about encroaching on some enforced hard character limit.

**Comment by ibuclaw** [comment]
Apart from the nits already mentioned (I may find more).  Has there been any benchmarking on this?  Are we loosing or gaining performance by switching to higher level accessors?  I haven't checked, but I assume that ``std.math`` is fully ``@nogc`` as well?

File: std/math.d (hunk 2 above)
Comment (ivan-timokhin) [inline std/math.d]: Both the docs at the top of the module and static asserts on lines 212-227 seem to imply that 80-bit `real`s are only supported on little-endian systems. Furthermore, AFAICT, all of the code currently dealing with those assumes little-endiannes (which is why I added `significand` in the first place).

**Comment by ivan-timokhin** [comment]
I didn't look into performance much (will do), but I'd expect the generated code to be more or less the same; e.g., in both https://goo.gl/umkAtu and https://goo.gl/jFzuR2 assembly output is the same for pointers and unions.

And yes, it seems that most functions are `@nogc`. `abs` is not marked as such, but has `@nogc` unittest; `approxEqual` does not have one (an oversight?); `resetIeeeFlags` has no attributes at all, but, again, that looks like an oversight.

File: std/math.d (hunk 2 above)
Comment (ibuclaw) [inline std/math.d]: Well, it's simple enough to reverse it for big-endian systems.

Comment (ivan-timokhin) [inline std/math.d]: Except that simply reversing the fields' order won't work because of alignment
issues, so I'll have to create an ad hoc version of Tuple here, or extend the
original one to allow specifying alignment. This is doable, but it is definitely
more work than I'm ready to devote to supporting a format for which I have no
simple and straightforward way of testing and which was not supported before,
especially since this PR is not supposed to add new functionality, being merely
a rewrite of already existing code.

Comment (ibuclaw) [inline std/math.d]: Is this alignment issues because of ``Tuple(ushort, ulong)`` or because of the accessor?  I would be very surprised if ``Tuple`` doesn't pack data, that would defeat the purpose of us using ``Tuple`` here in the first place.

Comment (ibuclaw) [inline std/math.d]: > This is doable, but it is definitely more work than I'm ready to devote to supporting a format for which I have no simple and straightforward way of testing

Yes, I know.  I was merely questioning the ``significand`` accessor for the purpose of porting to ``BigEndian`` - this is something that I will get around to, but not today.

Comment (ivan-timokhin) [inline std/math.d]: >I would be very surprised if `Tuple` doesn't pack data,

AFAICT, it doesn't. That's not a problem so far because the types are ordered appropriately, but might become one in the `BigEndian` case. The solution (if needed) is probably as simple as dropping `align(1):` in the `Tuple`'s body. This can be added as an option to existing `Tuple`, or as a small ad-hoc `private` type here; not sure which is better.

> >This is doable, but it is definitely more work than I'm ready to devote to supporting a format for which I have no simple and straightforward way of testing
>
>Yes, I know. I was merely questioning the `significand` accessor for the purpose of porting to `BigEndian` - this is something that I will get around to, but not today.

Come to think of it, that comment is a bit offensive. I'm sorry, didn't mean to.
Regarding `significand`, I only added it because there's a lot of code already in this module that manipulates the first `ulong` of `ieeeExtended` `real`s separately, so I thought that using a named accessor would look better and clearer than `integral[0]`. If that's causing so much trouble, I can drop it with no remorse.

Comment (ibuclaw) [inline std/math.d]: It's fine, because it makes the reverse format easier to manage also.  My only concern is with lack of packed align.

**Comment by ivan-timokhin** [comment]
:disappointed: Unfortunately, I must admit defeat. The resulting assembly does not look as similar as I hoped it would, and the new one is frequently longer. While for most functions it *seems* that the new instructions are mostly a slightly permuted version of old ones, my knowledge of assembly is not deep enough to tell if that's significant or not. 
An alternative is to run actual benchmarks on all module functions, but setting them up (as well as learning x86_64 assembly) would take more time and effort than I can dedicate at the moment.
So, in conclusion, I'm unable to provide any guarantees regarding performance of the new code compared to the old one.

**Comment by ibuclaw** [comment]
No problem. I can dig out a couple from my mailing list archives

**Comment by ibuclaw** [comment]
For the most part, I think this is fine.  Just having a mental pass at it, I can only see one notable place of concern.  Last nits will follow and I will give the green light

File: std/math.d [hunk 5]
```diff
@@ -5249,25 +5257,10 @@ bool isInfinity(X)(X x) @nogc @trusted pure nothrow
...
-    }
-    else
-    {
-        ushort* pxe = cast(ushort *)&x;
-        ushort* pye = cast(ushort *)&y;
-        return pxe[4] == pye[4] && pxs[0] == pys[0];
-    }
+    F.Layout repX = { number : x }, repY = { number : y };
+
+    return repX.integral == repY.integral;
```
Comment (ibuclaw) [inline std/math.d]: This would be turned into a ``memcmp`` for ``ieeeQuadruple`` and ``ibmExtended``, and goodness knows what for ``ieeeExtended``.  For efficiency, I'd recommend keeping the previous behaviour and inline the bit comparison.

File: std/math.d [hunk 6]
```diff
@@ -5690,58 +5683,58 @@ real nextUp(real x) @trusted pure nothrow @nogc
...
     }
-    return x;
+    return rep.number;
 }
 
 /** ditto */
 float nextUp(float x) @trusted pure nothrow @nogc
 {
-    uint *ps = cast(uint *)&x;
+    floatTraits!float.Layout rep = { number : x };
```
Comment (ibuclaw) [inline std/math.d]: Nit: ``floatTraits!(float).Layout``

File: std/math.d [hunk 7]
```diff
@@ -5690,58 +5683,58 @@ real nextUp(real x) @trusted pure nothrow @nogc
 /** ditto */
 double nextUp(double x) @trusted pure nothrow @nogc
 {
-    ulong *ps = cast(ulong *)&x;
+    floatTraits!double.Layout rep = { number : x };
```
Comment (ibuclaw) [inline std/math.d]: Nit: ``floatTraits!(double).Layout``

File: std/math.d [hunk 8]
```diff
@@ -6546,25 +6537,23 @@ body
     // average them (avoiding overflow), and cast the result back to a floating-point number.
 
     alias F = floatTraits!(T);
-    T u;
+    F.Layout repX = { number : x },
+                repY = { number : y },
+                res;
```
Comment (ibuclaw) [inline std/math.d]: Nit: Odd whitespace here.  I'd prefer keeping one declaration per statement.
```
F.Layout repX = {...};
F.Layout repY = {...};
F.Layout res;
```

File: std/math.d [hunk 9]
```diff
@@ -6577,60 +6566,50 @@ body
...
             m |= 0x4000_0000_0000_0000L; // shift carry into significand
         if (e)
-            *ul = m | 0x8000_0000_0000_0000L; // set implicit bit...
+            res.integral.significand = m | 0x8000_0000_0000_0000L; // set implicit bit...
         else
-            *ul = m; // ... unless exponent is 0 (subnormal or zero).
+            res.integral.significand = m; // ... unless exponent is 0 (subnormal or zero).
 
-        ue[4]= e | (xe[F.EXPPOS_SHORT]& 0x8000); // restore sign bit
+        res.shorts[4]= e | (repX.shorts[F.EXPPOS_SHORT]& 0x8000); // restore sign bit
```
Comment (ibuclaw) [inline std/math.d]: Nit: I can see that the previous code was this way, and you just exchanged ``ue`` and ``xe``, but...
```
res.shorts[4] = e | (repX.shorts[F.EXPPOS_SHORT] & 0x8000); // restore sign bit
             ^--- space added here, and also ---^
```

**Comment by ibuclaw** [comment]
Because you are dereferencing static arrays instead of pointers, I'd expect there to be more bounds checks, this will distort some things if you plan on doing comparison of assembly.

**Comment by ibuclaw** [comment]
> Why would indexing a static array by a (compile-time) constant offset cause a bounds check to be emitted?

All peeks are constant offsets, not all pokes are though.

**Comment by MartinNowak** [comment]
What's the state of this?

**Comment by ivan-timokhin** [comment]
I'm afraid I won't be able to devote any time or effort to this in a foreseeable
future, so if it doesn't look mergeable now, I'd suggest just closing this PR,
especially since it doesn't provide any tangible benefits to the end user right
now.

**Comment by ibuclaw** [comment]
Two optional nits, but otherwise I'm ok with this.
